enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,423 @@
+When the end user deletes the secret containing the certificate data, the `.spec.tls` on
+Route will be cleared. This will result in the route serving default certificates. In such
+a scenario the annotation that was added by the user will be automatically deleted by the
+secret-injector controller. This is only in the case where the user is directly creating
+routes instead of relying on the Ingresses.
+
+#### Variation [optional]
+
+N.A
+
+### API Extensions
+
+A new annotation of type `string` to be introduced as part of the integration
+to support third-party certificate management systems,
+
+```yaml
+annotations:
+  route.openshift.io/tls-secret-name: <secret-name>
+```
+
+_Note_: The default value would be N/A. The secret is required to be created
+in the same namespace as that of the Route. The secret must be of type
+`kubernetes.io/tls` and the tls.key and the tls.crt key must be provided in
+the `data` (or `stringData`) field of the Secret configuration.
+
+This annotation can be applied by the end user on Route. The controller that
+will be introduced as part of this enhancement will be responsible for the
+processing of this annotation on the Route.
+
+The Route API will also be updated to denote certificate status under `.status`,
+
+```go
+
+// RouteStatus provides relevant info about the status of a route, including which routers
+// acknowledge it.
+type RouteStatus struct {
+    // ...
+
+    // CertificateConditions describes the type of certificate that is being served
+    // by the router(s) associated with this route.
+    CertificateConditions []RouteCertificateCondition `json:"certificateConditions,omitempty" protobuf:"bytes,2,rep,name=certificateConditions"`
+}
+
+// RouteCertificateConditionType is a valid value forRouteCertificateCondition
+type RouteCertificateConditionType string
+
+// RouteCertificateCondition contains details of the certificate being served by routers associated with
+// this route.
+type RouteCertificateCondition struct {
+	// Type is the type of the condition.
+    // Possible values include Default, Custom and Managed.
+	Type RouteCertificateConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RouteCertificateConditionType"`
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
+	// (brief) reason for the condition's last transition, and is usually a machine and human
+	// readable constant
+	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+	// Human readable message indicating details about last transition.
+	Message string `json:"message,omitempty" protobuf:"bytes,4,opt,name=message"`
+	// RFC 3339 date and time when this condition last transitioned
+	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,5,opt,name=lastTransitionTime"`
+}
```

Comment:
I wonder if you're overcomplicating this with a condition? How much of this data do you actually need?

Seems to me like the controller could just denote its certificate status with an enum? 
```
certificateType: Default | Custom | Managed
```

Reply:
Hmm, that is a good suggestion. All three types were mutually exclusive so an Enum could work. Is there some sort of convention for `.status`, because I don't see a lot of API's deviating to have something new only in the `.status` (something not replicated from `.spec`). 

Apart from using `Reason`/`Message` to indicate why said type was evaluated by the controller and `status` to indicate the type in use currently no other fields are required.

Reply:
I think showing the user the `Message`, briefly explaining the "why" about the certificate type is helpful.

Reply:
I think we can switch to using `metav1.Condition`. The message for example would say something like the following,
```
// update status to Managed
cond := metav1.Condition{
    Type: ManagedCertificate,
    Status: metav1.ConditionTrue,
    Message: "TLS secret reference annotation is provided, certificate data is managed by an external entity.",
}

// update status to Default
// update message as defaulting due to secret not found
cond := metav1.Condition{
    Type: DefaultCertificate,
    Status: metav1.ConditionTrue,
    Message: "TLS secret reference annotation present but associated secret is missing, using default generated certificates.",
}

// update status to Custom
cond := metav1.Condition{
    Type: CustomCertificate,
    Status: metav1.ConditionTrue,
    Message: "TLS secret reference annotation missing, using custom certificate data since provided in .spec",
}
```
The important one here is the 2nd example as this indicates the scenario discussed [here](https://github.com/openshift/enhancements/pull/1307/files#r1091307533).

Reply:
That makes sense, lets go for `metav1.Condition` in this part then and keep it as in the examples you've provided